File: internal/audit/audit.go
// Package audit orchestrates read-only snapshot acquisition, parsing and analysis.
package audit

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"github.com/toddwbucy/Aletharsis/internal/analyzers"
	"github.com/toddwbucy/Aletharsis/internal/capability"
	"github.com/toddwbucy/Aletharsis/internal/evidence"
	"github.com/toddwbucy/Aletharsis/internal/failure"
	"github.com/toddwbucy/Aletharsis/internal/parsers"
	u "github.com/toddwbucy/Aletharsis/internal/unicoderef"
)

const Version = "0.2.0"
const MaxBytes = 8 * 1024 * 1024

type nativeCheck struct {
	id       string
	analyzer analyzers.Analyzer
}

var checks = []nativeCheck{
	{capability.UnicodeInventoryID, analyzers.Unicode{}},
	{capability.EmojiID, analyzers.Emoji{}},
	{capability.TextID, analyzers.Text{}},
	{capability.PatternsID, analyzers.Patterns{}},
	{capability.IdentifiersID, analyzers.Identifiers{}},
	{capability.MetadataID, analyzers.Metadata{}},
}

func Run(path string) *evidence.Report { return WithLimit(path, MaxBytes) }
func WithLimit(path string, limit int) *evidence.Report {
	return withReader(path, limit, readSnapshot)
}

// Outcome retains the typed native failure alongside an unchanged schema-1 report.
// It is an internal service result, not a new JSON report envelope.
type Outcome struct {
	Report  *evidence.Report
	Failure *failure.Error
}

// Inspect executes a native audit with explicit input bounds and retains its cause.
// No v2 fields are added to the legacy Report.
func Inspect(path string, limit int) Outcome {
	return inspectWithReader(path, limit, readSnapshot)
}

func withReader(path string, limit int, read func(string, int) ([]byte, error)) *evidence.Report {
	return inspectWithReader(path, limit, read).Report
}

func inspectWithReader(path string, limit int, read func(string, int) ([]byte, error)) Outcome {
	path = filepath.Clean(path)
	name := filepath.Base(path)
	if name == "." || name == string(filepath.Separator) {
		name = ""
	}
	extension := ""
	if i := strings.LastIndex(name, "."); i > 0 && i < len(name)-1 {
		extension = u.Lower(name[i:])
	}
	r := &evidence.Report{Version: Version, Schema: "1.0", File: evidence.File{Path: path, Filename: name, Extension: extension, MIME: "application/octet-stream", Format: "unknown", Basis: "unavailable"}, Status: "completed", Evidence: evidence.EmptyDocument(), Findings: []evidence.Finding{}, Limitations: analyzers.Limitations()}
	data, err := read(path, limit)
	if err != nil {
		typed := failure.AcquisitionError(err)
		return Outcome{Report: failed(r, typed), Failure: typed}
	}
	size, hash := len(data), evidence.Hash(data)
	r.File.Size = &size
	r.File.SHA256 = &hash
	r.File.Format, r.File.MIME, r.File.Basis = parsers.Identify(data, extension)
	if r.File.Format != "text" {
		typed := failure.Wrap(failure.Parsing, failure.UnsupportedFormat, fmt.Errorf("Unsupported format: %s; M0/M1 support Unicode text source files", r.File.Format))
		return Outcome{Report: failed(r, typed), Failure: typed}
	}
	parser := parsers.TextParser{}
	parserName := parser.Name()
	r.File.Parser = &parserName
	r.Evidence, err = parser.Parse(data)
	if err != nil {
		code := failure.AuditFailed
		var decode *parsers.DecodeError
		if errors.As(err, &decode) {
			code = failure.DecodeFailed
		}
		typed := failure.Wrap(failure.Parsing, code, err)
		return Outcome{Report: failed(r, typed), Failure: typed}
	}
	for _, check := range checks {
		r.Findings = append(r.Findings, check.analyzer.Analyze(&r.Evidence)...)
	}
	sort.SliceStable(r.Findings, func(i, j int) bool {
		a, b := r.Findings[i], r.Findings[j]
		rank := func(s string) int {
			if s == "INFO" {
				return 0
			}
			return evidence.Rank(s)
		}
		if rank(a.Severity) != rank(b.Severity) {
			return rank(a.Severity) > rank(b.Severity)
		}
		if a.ID != b.ID {
			return a.ID < b.ID
		}
		ak, bk := locationKey(a.Location), locationKey(b.Location)
		if ak != bk {
			return ak < bk
		}
		return a.Title < b.Title
	})
	r.Summarize()
	return Outcome{Report: r}
}
func locationKey(loc evidence.Object) string { b, _ := json.Marshal(loc); return string(b) }
func failed(r *evidence.Report, err error) *evidence.Report {
	r.Status = "failed"
	kind := "ValueError"
	details := evidence.Object{}
	var decode *parsers.DecodeError
	if errors.As(err, &decode) {
		kind = "UnicodeDecodeError"
		details["byte_start"] = decode.Start
		details["byte_end"] = decode.End
		details["reason"] = decode.Reason
	} else if errors.Is(err, os.ErrNotExist) {
		kind = "FileNotFoundError"
	} else if errors.Is(err, os.ErrPermission) {
		kind = "PermissionError"
	} else {
		var pe *os.PathError
		if errors.As(err, &pe) {
			kind = "OSError"
		}
	}
	details["error_type"] = kind
	details["message"] = err.Error()
	r.Findings = append(r.Findings, evidence.Finding{ID: "parser.failure", Severity: "INFO", Confidence: 1, Category: "parser", Classification: "observed_fact", Title: "Audit could not complete", Description: "The file was not fully analyzed. Review the error before interpreting findings.", Evidence: details, Location: evidence.Object{}})
	r.Summarize()
	return r
}

File: src/aletharsis/audit.py
"""Read a bounded snapshot, parse it, then run independent analyzers."""

import hashlib
import os
from pathlib import Path
import stat

from . import __version__
from .analyzers.base import Analyzer
from .analyzers.identifiers import IdentifierAnalyzer
from .analyzers.metadata import MetadataAnalyzer
from .analyzers.patterns import PatternAnalyzer
from .analyzers.text import TextAnalyzer
from .analyzers.unicode import UnicodeAnalyzer
from .identify import identify
from .models import FileIdentity, Finding, Report
from .parsers.base import Parser
from .parsers.text import TextParser

MAX_BYTES = 8 * 1024 * 1024
PARSERS: dict[str, Parser] = {"text": TextParser()}
ANALYZERS: tuple[Analyzer, ...] = (UnicodeAnalyzer(), TextAnalyzer(), PatternAnalyzer(),
                                  IdentifierAnalyzer(), MetadataAnalyzer())
LIMITATIONS = [
    "A suspicious artifact is not necessarily a watermark. Observable evidence and structural patterns do not establish intent or provenance.",
    "Statistical model-output watermarks cannot be determined by these analyzers. No AI authorship or vendor attribution is performed.",
    "M0/M1 inspect literal text source only: HTML/XML markup, JSON escapes, entities and rendered content are not decoded or interpreted. Structured metadata is not extracted.",
    "DOCX/PDF and batch auditing are not implemented in this milestone. Absence of findings is not proof of absence of a watermark.",
    "Pattern thresholds are heuristics, not calibrated probabilities. Confidence describes support for the stated observation or pattern, not watermark intent.",
]


def read_snapshot(path: Path, max_bytes: int) -> bytes:
    if not hasattr(os, "O_NOATIME"):
        raise OSError("This platform lacks O_NOATIME; strict timestamp-preserving reads are unavailable")
    descriptor = os.open(path, os.O_RDONLY | os.O_NOATIME | os.O_NOFOLLOW | os.O_NONBLOCK)
    with os.fdopen(descriptor, "rb") as stream:
        before = os.fstat(stream.fileno())
        if not stat.S_ISREG(before.st_mode):
            raise ValueError("Only regular files are supported; directory auditing is planned for M4")
        if before.st_size > max_bytes:
            raise ValueError(f"File exceeds the {max_bytes}-byte analysis limit")
        data = stream.read(max_bytes + 1)
        after = os.fstat(stream.fileno())
    if len(data) > max_bytes:
        raise ValueError(f"File exceeds the {max_bytes}-byte analysis limit")
    if (before.st_size, before.st_mtime_ns, before.st_ctime_ns) != (after.st_size, after.st_mtime_ns, after.st_ctime_ns):
        raise ValueError("Source changed during snapshot acquisition; evidence may be inconsistent")
    return data


def audit(path: Path, *, max_bytes: int = MAX_BYTES) -> Report:
    path = Path(path)
    identity = FileIdentity(str(path), path.name, path.suffix.lower())
    report = Report(__version__, "1.0", identity, limitations=list(LIMITATIONS))
    try:
        data = read_snapshot(path, max_bytes)
        identity.size = len(data)
        identity.sha256 = hashlib.sha256(data).hexdigest()
        identity.format, identity.mime, identity.identification_basis = identify(data, path)
        parser = PARSERS.get(identity.format)
        if parser is None:
            raise ValueError(f"Unsupported format: {identity.format}; M0/M1 support Unicode text source files")
        identity.parser = parser.name
        report.evidence = parser.parse(data)
        for analyzer in ANALYZERS:
            report.findings.extend(analyzer.analyze(report.evidence))
        report.findings.sort(key=lambda f: (-{"INFO": 0, "LOW": 1, "MEDIUM": 2, "HIGH": 3}[f.severity], f.id,
                                            str(f.location), f.title))
    except (OSError, ValueError) as exc:
        report.status = "failed"
        details: dict = {"error_type": type(exc).__name__, "message": str(exc)}
        if isinstance(exc, UnicodeDecodeError):
            details.update(byte_start=exc.start, byte_end=exc.end, reason=exc.reason)
        report.findings.append(Finding(
            "parser.failure", "INFO", 1.0, "parser", "observed_fact", "Audit could not complete",
            "The file was not fully analyzed. Review the error before interpreting findings.", details))
    return report
